src/rules/prefer-readonly/classScope.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+const OUTSIDE_CONSTRUCTOR = -1;
+const DIRECTLY_INSIDE_CONSTRUCTOR = OUTSIDE_CONSTRUCTOR + 1;
+
+interface VariableModifications {
+    toMember?: boolean;
+    toStatic?: boolean;
+}
+
+export type ParameterOrPropertyDeclaration = ts.ParameterDeclaration | ts.PropertyDeclaration;
+
+function typeIsOrHasBaseType(type: ts.Type, parentType: ts.Type) {
+    if (type.symbol === undefined || parentType.symbol === undefined) {
+        return false;
+    }
+
+    for (const baseType of [type, ...type.getBaseTypes()]) {
+        if (baseType.symbol !== undefined && baseType.symbol.name === parentType.symbol.name) {
```

Comment:
That's probably ok for now.
There's the internal typescript function `isTypeInstanceOf` in `checker.ts` that does exactly what we need here. We should ask to expose this functionality.